### ISSUE TITLE
Ingest SST files to L0 

### ISSUE THREAD

**Issue by patelprateek** [facebook/rocksdb#11671]
Is there a way to  ingest SSt files  Level  L0 only , is there a way to enforce that newly added SST files go to L0 level only ?
I have a use case where I do all writes on indexing machine and do full compaction at the end and move all data (along with mainifest to another machine whic operate on read only query mode .

This works well for batch mode , but i have streaming data coming in and i want to push updates faster to read only rocksdb instance , when my index grows large like few hundred GB doing compaction and moving data to another instance adds significant delays and I was wondering if I can directly move the smaller delta L0 files only out of my write only rocksDB instance to L0 level on read only instance and periodically do compaction on write only side. ?

Any recommendation on how to achieve this

**Comment by ajkr** [comment]
> is there a way to enforce that newly added SST files go to L0 level only ?

No, it is typically undesirable for newly added SST files to go to L0 level since some parts of the code (for example, write stalling logic) treat L0 files as though they span the whole key-range, even if they don't. Besides that, it isn't guaranteed you can take an L0 file from one DB and ingest it to another DB. Files for ingestion need to be created by `SstFileWriter`, while L0 files could also be created by memtable flush.

> Any recommendation on how to achieve this

Given you proposed a solution involving ingesting to L0, I am assuming that your streaming updates are through an ingestion API (i.e., `SstFileWriter`+`IngestExternalFile()`) and not through a live write API (e.g., `Put()`). In that case, can you instead send files written by `SstFileWriter` from the write-only instance to the read-only instances? Then do the ingestion on the read-only side, which might require temporarily opening it in read/write mode. You can also periodically ingest those same files on the write-only side, re-compact, and redistribute the fully compacted DB.

**Comment by patelprateek** [comment]
Thanks for the suggestions 

Some more context 
Streaming updates are coming through messaging queue (kafka/pubsub) .
The reason i was going for L0 files are because , they indeed span entire range (i dont have control on streaming data generation)  , so i thought its suitable for them to be ingested at L0 in the read only instance. 

Is there a difference in L0 SST files vs SSTWriter created SST files  ?  Since i have unbounded streaming data , instead of sorting it myself and creating sst files , i instead write to memtable and let flush operation create the SST files for me ? 
Otherwise doing sorting in multiple threads will create huge number of files (i would have to tune myself the batch size , frequency etc etc) , so instead i let all threads dump data to large enough memtables and let it flush based on L0 bytes or num of files trigger and then compact if needed before sending it over to read only instance (that was my plan) . Basically SSTFile writer needs sorted data which requires upstream data generation to provide me sorted data or my ingestion code create such windowed batch and do sorting for each thread , so instead i kind of re-used rocksdb to do the sorting for me using memtable merge + flush + compact operations. 

So to clarify :  ingestion (write-only instance on indexing vm)  subscribes to a messaging queue , uses the Live api to generate sst files (using memtable flush) , after compactions, copy data over to query (read-only) machines.  This works fine for batch , but need to scale it for faster updates for example i have about ~1GB per minute updates , that i want to ideally be searchable on query nodes as soon as possible . If the base index itself is 200 GB , periodically compacting this update with the base index and sending ~200GB snapshots within say 30 seconds is hard and will not scale if update data size or base index grows (may be 2x-5x in future , unless i horizontally scale to more shards).   So Instead as you suggested i was planning to send directly the small L0 sst files to query only node , and then periodically run a full compaction to send over the entire 200 GB (otherwise too many sst files will degrade serving read query performance due to too many sst files)

**Comment by ajkr** [comment]
Oh ok, thanks for the info - I didn't realize the streaming update is using live write APIs so need to backtrack a bit.

There actually is a way to ensure flushed memtables remain in L0, which is to simply set `disable_auto_compaction=true`. However the files it produces still can't be ingested to another DB.

> Is there a difference in L0 SST files vs SSTWriter created SST files ?

It's not a huge difference, but there are some differences in the table properties that are written by flush/compaction vs. expected by ingestion, so currently it fails. I tried once to make all files ingestible (https://github.com/ajkr/rocksdb/commit/bc43e47b31adbe24a6ec97054680c9450ccbf22c), and really it wasn't that hard. Maybe we should finish that some day.

On the other hand, it is error-prone to allow users to ingest any file from one DB to another. We don't have the ability to validate, for example, that you really are ingesting successively newer flushed files from the source DB.

Aside from that, have you looked into our secondary instance support (https://github.com/facebook/rocksdb/wiki/Read-only-and-Secondary-instances)? I wonder if your incremental update workflow can be like:

(1) `disable_auto_compaction=true` on write-only instance
(2) create L0 files through live update and memtable flush
(3) copy MANIFEST and new L0 files over to secondary instances
(4) call `TryCatchupWithPrimary()` on the secondary instances

**Comment by patelprateek** [comment]
Thanks for the suggestions 
yes i disable auto compaction to ensure all memtables are being flushed to L0

I did read a bit about secondary instances a bit but there are few things that led me to a different way.

I thought `TryCatchupWithPrimary` works with a remote shared attached disk where the primary instance keeps adding new sst files and updating manifest and a separate process on a separate remote VM uses  secondary instance to open this same shared attached disk . The remote SSD read IOPS would not be sufficient to high QPS and latency requirements on the serving side and hence i was looking for other workarounds

IIUC , what you are suggesting is instead of using network attached disks , copy files (new L0 sst and manifest) to remote machines local ssd and then call `TryCatchupWithPrimary` on the local SSDs copied data using secondary instance ? This seems interesting . 1) how to figure out new SST (can this be inferred from manifest file) ? 2) too many L0 SST file can impact read latency so if i do periodic compaction on the write only instance will this strategy still work (i would need to copy over the compacted L1 .. Ln ) files as well in this case  ?  

One concern i have was since its streaming data , eventually the data can grow up to few hundred gigabytes , so copying over all data for example 200 GB over to read-only instances every 30 seconds or less will be hard .  Hence i only wanted to copy new SST files and perhaps copy all SST files (all levels) only if i run a compaction on the write-only side (want to avoid copying large SST files that are at lower levels).
 Is there a quick way to identify what SST files have changed across all levels to make this copying process faster, i.e sync only the files that have changed

**Comment by patelprateek** [comment]
Thanks for the detailed explanation. 

Few more questions : 
1) is there a plan to make all files ingestable (sstwriter and liveapi ) ? 
2) what is the typical use case for SSTWriter , if the writer expects sorted input it seems to be very limited , since the job doing the sorting needs to do almost the same work that the live api is doing : sort + flush + merge  to handle large data sets . the live api gives all the controls for file size , where as doing sorting on an map reduce job one needs to re-write a lot of same thing.  If i decide to use live api , can i generate an ingestable file post final compaction i.e iterate over my DB and generate ingestable file , this way i dont need to path rocksDB source code and can keep using the maven available jni

**Comment by ajkr** [comment]
> is there a plan to make all files ingestable (sstwriter and liveapi ) ?

It might happen one day when we run out of available options for a particular use case, but it isn't something I'd want to accelerate. The main thing I worry about is API safety since we have no way to enforce that files are ingested on the destination in the correct order.

Although, we do have use cases for which this isn't a valid concern as RocksDB does not implement their ordering and they do not use snapshots. For example, a use case where all operations are merges and the merge operator provides the ordering, or a use case where the user key contains version info making every key unique. In those cases RocksDB's internal sequence number is extraneous as it does not order anything.

For such use cases, though, I would prefer we first make it explicit via some API that ordering is not provided by RocksDB. This would be convenient for other reasons, like we wouldn't need to store sequence numbers for them, saving space and memory. We could then allow arbitrary ingestion only for those use cases.

In terms of safe APIs that are already available, I prefer `GetLiveFilesStorageInfo()` with either `disable_auto_compaction=true` or `EnableFileDeletions()` on the primary, and `TryCatchupWithPrimary()` on a secondary or `Close()`/`Open()` on a read-only. That way your secondary/read-only instances get the exact manifest entries that the primary wrote. We could probably find ways to make these APIs more convenient for determining delta of files or making secondary work better on non-shared storage.

Will get to your second question later as I have a bit to say there too...

**Comment by ajkr** [comment]
>  If i decide to use live api , can i generate an ingestable file post final compaction i.e iterate over my DB and generate ingestable file , this way i dont need to patch rocksDB source code and can keep using the maven available jni

For transferring the full DB, yes. Post final compaction or not is irrelevant since either way the DB iterator will only return the latest version of each key.

For transferring incremental changes, it depends. If your updates are all insertions and overwrites, then yes. But if there are deletions, then no, since a DB iterator does not see deletion markers.

> the live api gives all the controls for file size , where as doing sorting on an map reduce job one needs to re-write a lot of same thing.

This is true and we were already thinking about adding the file cutting logic to `SstFileWriter` to help with this. The change I am thinking about is to start allowing the user to provide a directory in the `file_path` argument - https://github.com/facebook/rocksdb/blob/fb5748decfbe7d42f9a4dfd4925b5bc61bb2ea45/include/rocksdb/sst_file_writer.h#L110C45-L110C45

When a directory is provided, RocksDB would write files to that directory following a naming convention of our choice. We would cut files according to the target sizes that `SstFileWriter` already knows from the user's settings. You would find out the files by listing the directory

**Comment by patelprateek** [comment]
One quick clarification , if i do use SSTFilewriter , for a streaming source , when sorting it myself i do need a stable sorting right ? because my events are ordered , sorting them should not re-order the order of events for a given key (i think the live api memtables handle this seamlessly for user but for sstwriter we need to take care of it)

The other thing is when ingesting sorted files , how can i enable ttl for data within it, it seems to me the entire file will be ingested with one time stamp ? 

My current approach : 
- take streaming data and every few seconds (window) generate ingestable SST file (preferably with insertion timestamp for enabling ttl expiry later) : lets call this delta tables . Not sure if i should use live api or sstwriter yet.
- the small ingestable SST files will  be transferred to read-only DB (this should be in order of few seconds due to small amount of data so flush + compaction + moving files to remote storage and copying back to local ssd on read-only should be fast ) 
- Have another compaction job that takes a large index (base snapshot periodically generated by offline batch process) . Then ingest these delta tables in the same order as they are produced : do you know if we ingest external sst files , can we enable ordering of ingestion of these files ? During this compaction process (base + several delta files ). i will try to delete ttl expired data as well , this large compacted DB can be moved to read-only instance every couple minutes 
Basically  I am trying to apply few hundred MB updates every few seconds to a large (few hundred gb DB) ,and make them searchable on read side as soon as possible , just running the live api insertion on large compacted db and moving few hundred gb every few seconds will not be scalable and hence trying to move small deltas  and on.seperate process doing comapaction and updating the read-only db with a lesser cadence so that read perf is not deterioted due to too many small detla tables. Let me know if you think this approach works (NOTE : we do need to support ttl expiry else the data would be ever growing since its streaming)

**Comment by ajkr** [comment]
> when sorting it myself i do need a stable sorting right ? because my events are ordered , sorting them should not re-order the order of events for a given key (i think the live api memtables handle this seamlessly for user but for sstwriter we need to take care of it)

The `SstFileWriter` API does not support duplicate key versions. If you only retain latest version of a given key, then your sort doesn't need to worry about stability. If you want to store multiple versions of a given key, you'll need to append version info like timestamp to the key to turn them into distinct keys.

> The other thing is when ingesting sorted files , how can i enable ttl for data within it, it seems to me the entire file will be ingested with one time stamp ?

Every key in the file will be assigned the same sequence number. What timestamp do you plan to use for expiration? For supporting variable per key TTLs you should store the expiration timestamp not the event timestamp. For supporting historical reads you should include the event timestamp in the key not the value, either using our user-defined timestamp or appending the event timestamp to the key.

> can we enable ordering of ingestion of these files ? During this compaction process (base + several delta files ). i will try to delete ttl expired data as well , this large compacted DB can be moved to read-only instance every couple minutes

The `SstFileWriter` and `IngestExternalFile()` approach will be problematic here unless you create a new DB. That's because ingestion will write all the files as newer than all existing ones, which could land all your ingested files in L0, slowing reads further.

**Comment by patelprateek** [comment]
Perhaps i am missing some information , why would i need to create new DB  (i do want all new ingested files to be newer and ingested in same order as they are produced) ? 
to summarize the idea was 
- sstfilewriter (or live api) => generate small sorted files (for ex : 64 mb every 15 seconds) : delta_0 (Ts:0), delta_1 (Ts:15 ), delta_2 (Ts:30) ...
- Enable moving these delta to read only DB and ingest them in same order i.e ingest delta_0 , then delta_1  , delta_2 (I assume ingestion can be ordered and all these deltas land in L0 of read only DB , are you suggesting when ingesting multiple deltas, we no longer know whether data in delta_0 is older than delta_1 )
- Since lot of deltas will slow read perf , the write only periodically generates (base (100 GB)  + delta_0 (64 mb) +delta_1 .. : compacted file ~100 GB ) every 10 minutes and moves it to read only DB , where read-only DB opens a new DB pointing to the new compacted set of files

**Comment by ajkr** [comment]
> Enable moving these delta to read only DB and ingest them in same order i.e ingest delta_0 , then delta_1 , delta_2 (I assume ingestion can be ordered and all these deltas land in L0 of read only DB , are you suggesting when ingesting multiple deltas, we no longer know whether data in delta_0 is older than delta_1 )

This part is fine. delta_0 older than delta_1 is preserved when you ingest them in the same order. You don't need to create a new DB here.

> Since lot of deltas will slow read perf , the write only periodically generates (base (100 GB) + delta_0 (64 mb) +delta_1 .. : compacted file ~100 GB ) every 10 minutes and moves it to read only DB , where read-only DB opens a new DB pointing to the new compacted set of files

Yes this is the new DB. It sounds like you were already planning to open a new DB based on the fully compacted DB, which is the same as what I meant to suggest.
